### DISCUSSION TITLE
Business logic - what do we think about having a 'dry run' type thing in the cc-node to give feedback before trying a particular transaction?

### DISCUSSION THREAD

**Discussion by utunga** [CredComSoc/ProtocolWorkingGroup#16] [Client Applications]
I was looking over [*this* PR](https://github.com/CredComSoc/CreditCoopTech/pull/187/files) and saw this comment 

![image](https://github.com/CredComSoc/ProtocolWorkingGroup/assets/166867/c665c735-3f84-4fb6-93bc-aab042a3f348)

With mutual credit systems the logic of positive and negative limits can be a bit confusing. 

I think it should be that the logic -  deciding if a transaction is or is not valid - should be in the protocol.

- Should be in the CC-Node layer, first and foremost  (Maybe it is already)
- Secondly, if the CC-Node layer enforces such logic it then it's *OK* if the logic is implemented  in the UX layer **again** -  so as to give a more friendly experience so a user is warned 'that would exceed your balance' type thing even before they hit submit. 

**But here's my question.**  Does it make sense to have a thing where the API can sort of run a dry-run of a particular transaction and provide some feedback? 

To help people implementing UX against the interface - would it be useful (and worth it?) to have an end point that gives feedback of what *would* happen if the user was to attempt to try this transaction (assuming no other changes happen in the meantime).

I guess to retain i18n ability it might have to return a bunch of op_codes, instead of actual text - things like '`TRANSACTION_OK`', '`PAYEE_POS_BALANCE_LIMIT`' '`PAYER_NEG_BALANCE_LIMIT`' etc but y'know something in the form of helpful responses that the implementer can then wrap an interface around?

What do people think? @matslats , @JohnG-cc @yonasbek ?

Also is it OK and appropriate for me to start this conversation here, and in the Version 1 forum?

**Comment by yonasbek** [comment]
@utunga You are right to suggest that the logics of that should be handled from the cc-node with some kind of error codes like you mentioned above. And I think there is an implementation of that on the current cc-node but it is kind of vague and not clear. And tom wododrof's take on how to display the error message differs from one violation to another and I think I am not sure but the current implementation only have two violations but in the front end we check may be for six or may be five violations.

**Reply by tfwoodroofMCS** [reply]
If it's helpful here's a list of violations and corresponding error messages:

1. Purchase request that would violate buyer's negative limit: **Pop-up message to buyer**: 'You do not have enough credit to request this purchase, as the total price is `total price` `credit unit` and your available credit is `available credit` `credit unit`.' (https://github.com/CredComSoc/CreditCoopTech/issues/91). [Implemented].

2. Purchase request that would violate seller's positive limit: **Pop-up message to buyer**: 'This purchase request cannot be completed, as it would take `seller username 1`, `seller username 2` and `seller username 3` beyond their positive credit limit(s). These seller(s) have been notified, but please consider following up with them about your intended purchase.' (https://github.com/CredComSoc/CreditCoopTech/issues/86). **Notification to seller(s)**: '`buyer username` initiated a purchase request with you for `total price` `credit unit`, but this could not be processed as it would have taken you beyond your positive credit limit by `amount` `credit unit`. Please consider following up with `buyer username` about their intended purchase.' (https://github.com/CredComSoc/CreditCoopTech/issues/88). [Implemented].

3. Purchase request that would violate both of the above: **Pop-up message to buyer**: 'You do not have enough credit to request this purchase, as the total price is `total price` `credit unit` and your available credit is `available credit` `credit unit`. /linebreak Additionally, this purchase request would take `seller username 1`, `seller username 2` and `seller username 3` beyond their positive credit limit(s). These seller(s) have been notified, but please consider following up with them about your intended purchase.' (https://github.com/CredComSoc/CreditCoopTech/issues/90) [Not currently implemented]. **Notification to seller(s)**: '`buyer username` initiated a purchase request with you for `total price` `credit unit`, but this could not be processed as it would have taken you beyond your positive credit limit by `amount` `credit unit`. Please consider following up with `buyer username` about their intended purchase.' (https://github.com/CredComSoc/CreditCoopTech/issues/89) [Not currently implemented].

4. Seller approval of purchase request that would violate buyer's negative limit: **Pop-up message to seller**: 'The purchase request cannot be approved, as it would take `buyer username` below their negative credit limit. Please consider following up with them.' (https://github.com/CredComSoc/CreditCoopTech/issues/126). [Implemented as a failsafe (shouldn't be possible for buyers to make such purchase requests in the first place), but needs some testing].

5. Seller approval of purchase request that would violate seller's positive limit: **Pop-up message to seller**: 'The purchase request cannot be approved, as your balance would then exceed your positive credit limit by `excess credit` `credit unit`.' (https://github.com/CredComSoc/CreditCoopTech/issues/103) [Implemented, but needs testing once https://github.com/CredComSoc/CreditCoopTech/issues/165 has been implemented to allow this condition to be possible in the first place].

**Reply by utunga** [reply]
Very comprehensive list @tfwoodroofMCS I appreciate it (though i feel a tiny bit guilty for triggering that amount of work).

Hmmm..🤔

**Comment by JohnG-cc** [comment]
This seems like the right place for the discussion, but yes it could go under one of the other Track 1 categories instead if that seems better.

The protocol and reference implementation already have a couple of features that seem adjacent to what you propose, namely the fact that a transaction is "tested" across the network to calculate fees, and that the sender is then able to confirm the total with fees or cancel. I imagine this is probably the only way to really give an accurate assessment of the transaction outcome ahead of time. BTW this also places a hold on the required amount of clearing account balances on each node, so it should not be done flippantly.
Also, the reference implementation has plugin points for a business rule engine - it seems like this would be the right place to put any checks above and beyond the existing "sufficient balance" checks?

I don't know the implementation well enough to really know if this can give you what you need though. @mattslats would need to confirm that.

**Comment by matslats** [comment]
> Should be in the CC-Node layer, first and foremost (Maybe it is already) 
The cc-node layer DOES enforce this logic 

>  if the CC-Node layer enforces such logic it then it's OK if the logic is implemented in the UX layer again - so as to give a more friendly experience so a user is warned 'that would exceed your balance' type thing even before they hit submit.
The protocol defines a precise language-neutral error object indication the current limit and by how much the transaction would exceed the limit. You can get this error message before hitting submit by doing an ajax submission, but I would advise querying the balance limits and setting a max on the numeric field.

> To retain i18n ability it might have to return a bunch of op_codes like
It already does much more than this.

> it is kind of vague and not clear.
It should be stated in the API doc with whatever clarity that entails: Error class MaxLimitException and MinLimitException. See https://gitlab.com/credit-commons/cc-php-lib/-/blob/master/src/Exceptions/MaxLimitViolation.php

> Also, the reference implementation has plugin points for a business rule engine - it seems like this would be the right place to put any checks above and beyond the existing "sufficient balance" checks?
No the Blogic module can only return new Entries. It cannot validate the transaction. Currently there is no way to perform validation additional to balance limits in the reference implementation.
